### ISSUE TITLE
Slow performance of the tags

### ISSUE THREAD

**Issue by cyankalium** [tagspaces/tagspaces#1176]
Hello,

I use TagSpacesPro 3.5.4, I have 6 tag groups which consist of ~10-15 tags each, so in sum ~70 different tags. When I browse the folder of my tagged files (each file has ~10 tags and the folder has ~1000 files) its extremely laggy, although my PC specs (16 GB RAM, 4.5 Ghz Quadcore, Win 10 Pro 2004) should be fine.

If I disable that the tags are shown in the file explorer the lag is almost gone. But I wan't to edit tags and browse files, is it possible to build some sort of index or database for the tags so the performance issue is gone (1000 files à 10 Tags is not that much).?

This would be greatly appreciated.

**Comment by gitdocke** [comment]
I support this, I have the same problem and Tagspaces is showing its limits.
At this time I have indexed 3500 files in more than 100 directories plus 2700 different tags and the performance is going down by the day because of new additions. 
The autocomplete feature when I type tags is fundamental, however at this point is not possible to use it because of Tagspaces' slowing down.
I request some sort of SQL to get instant searches and snappy behavior. 
I know this wishful endeavor is not easy to implement, and even more when interoperability through different platforms is considered.

**Comment by cyankalium** [comment]
SQL would be awesome indeed. I assume storing all tags / json / sidecar files into the RAM is the way it's currently implemented?

**Comment by uggrock** [comment]
the problem here is that the app tries to show all the files at once. We will address issue by adding a pagination for lets say over 100 elements. In general if the 1000 files are distributed in sub folders with up to 400 files each, TagSpaces has in general no performance issues.

**Comment by gitdocke** [comment]
@uggrock 
Yes, I followed your advise a long time ago and I created many subfolders to make easier to load the content (or file thumbnails), and when I did it the improvement in performance was immediate.

However, as I was adding more files and tagging them heavily (some files have like 40 tags), I hit a new performance bottleneck, and the workaround to get a functional Tagspaces again was to disable the collection of tags and to delete most tag groups. This was like more than a year ago.

So, after that, more files were added and tagged, the new slow down now is about to drag & drop tags among files, the lag is growing by the day and I'm afraid soon will not be able to use Tagspaces any longer.

**Comment by sytolk** [comment]
@cyankalium @gitdocke can you check the performance with pagination here: https://github.com/tagspaces/tagspaces/pull/1183

p.s. generally you are right FileStructure sorting and so on are for the back-end task like Relational or NoSQL DB but I dont think that MySQL is the right choose. SQLite it is embeddable maybe.. but I'm thinking about couchbase.

**Comment by gitdocke** [comment]
@sytolk
For safety reasons, as long as Tagspaces stays isolated from the network, it will be fine.

**Comment by gitdocke** [comment]
@uggrock @sytolk 
Reporting.
I just gave a first run to v3.6.2 community version and I'm impressed.
As long I use Tagspaces without tag groups, directories will load almost in a blink and searches are a pleasure. Great improvement here.

Now the not so good part:
So, I decided to put Tagspaces under stress. Using v3.1.4 Pro, which allows me to collect all the tags without the 500 tags restriction (I've got 2950 now) I tested Tagspaces v3.6.2 again and I'm partially pleased. 
The auto complete feature is loading the tag's list at light speed, directories with a lot of huge files have a small delay loading but it's fine. Dragging and dropping tags among files also improved greatly, before there was a delay of 2 seconds to make the tag to stick, now perhaps 0.4 second.
However, the search is incredibly slow, looking for a tag will show results after 15 to 20 seconds.
In a nutshell, enabling full tag collection in v3.6.2 the search feature is unbearable.

![1](https://user-images.githubusercontent.com/22734449/93428676-e1c3bc80-f88d-11ea-8afa-b9271380aa84.png) 
![2](https://user-images.githubusercontent.com/22734449/93428705-eb4d2480-f88d-11ea-90d5-429fa8ad2ccb.png)
